src/clientManager.js

```diff
@@ -36,135 +36,233 @@ ClientManager.prototype.findClient = function(name, token) {
+	let self = data.this || this;
+	let loadusers = data.users || [];
+	return new Promise((resolve, reject) => {
+		// let current = input.users
+		// load users what are missing
 
-ClientManager.prototype.addUser = function(name, password, enableLog) {
-	var users = this.getUsers();
-	if (users.indexOf(name) !== -1) {
-		return false;
-	}
-	try {
-		if (require("path").basename(name) !== name) {
-			throw new Error(name + " is an invalid username.");
-		}
+		const current = self.clients.map(c => c.name);
+		let loadpromises = _.difference(loadusers, current)
+			.map((user) => self.loadUserPromise({name: user}));
 
-		var user = {
-			user: name,
-			password: password || "",
-			log: enableLog,
-			networks: []
-		};
-		fs.writeFileSync(
-			Helper.getUserConfigPath(name),
-			JSON.stringify(user, null, "\t")
-		);
-	} catch (e) {
-		log.error("Failed to add user " + name, e);
-		throw e;
-	}
-	return true;
+		Promise.all(loadpromises)
+			.then(() => resolve(data))
+			.catch((err)=> {
+				reject(err);
+			});
+	});
 };
 
-ClientManager.prototype.updateUser = function(name, opts, callback) {
-	const users = this.getUsers();
-	if (users.indexOf(name) === -1) {
-		return false;
-	}
-	if (typeof opts === "undefined") {
-		return false;
-	}
-
-	let user = this.readUserConfig(name);
-	const currentUser = JSON.stringify(user, null, "\t");
-	_.assign(user, opts);
-	const newUser = JSON.stringify(user, null, "\t");
-
-	// Do not touch the disk if object has not changed
-	if (currentUser === newUser) {
-		return callback ? callback() : true;
-	}
+ClientManager.prototype.addUserPromise = function(data) {
+	let self = data.this || this;
+	let name = data.name || undefined;
+	let password = data.password || undefined;
+	let enableLog = data.enableLog || undefined;
+	return new Promise((resolve, reject) => {
+		self.getUsersPromise()
+			.then((input) => {
+				let users = input.users;
+				if (users.indexOf(name) !== -1) {
+					reject(false);
+					throw new Error("User " + name + " already exist.");
+				}
+				if (require("path").basename(name) !== name) {
+					reject(false);
+					throw new Error(name + " is an invalid username.");
+				}
+				const user = {
+					user: name,
+					password: password || "",
+					log: enableLog,
+					networks: []
+				};
+				fs.writeFile(Helper.getUserConfigPath(name),
+					JSON.stringify(user, null, "\t"),
+					"utf-8",
+					(err) => {
+						if (err) {
+							reject(err);
+						} else {
+							resolve(data);
+						}
+					});
+			})
+			.catch((err) => {
+				log.error(`Error adding user ${colors.bold(name)}.`, err.message);
+			});
+	});
+};
 
-	fs.writeFile(Helper.getUserConfigPath(name), newUser, (err) => {
-		if (err) {
-			log.error("Failed to update user", err);
-		}
+ClientManager.prototype.updateUserPromise = function(data) {
+	let self = data.this || this;
+	let name = data.name || undefined;
+	let opts = data.opts || undefined;
+	return new Promise((resolve, reject) => {
+		self.getUsersPromise()
```

Comment:
Get rid of this `getUsers`, all calls to `updateUserPromise` seem to already check for user existance.